Permanent/ViewControllers/TagDetailsViewController.swift

```diff
@@ -0,0 +1,260 @@
+        dismiss(animated: true, completion: nil)
+    }
+    
+    @objc func doneButtonPressed(_ sender: UIBarButtonItem) {
+        let dispatchGroup = DispatchGroup()
+        showSpinner(colored: .lightGray)
+        
+        let forAddingBool = sortedArray.map({ (SortedTagVO) in
+            return SortedTagVO.forAdding
+         })
+        let forAddingNames = sortedArray.map { (item) -> String in
+            return item.tagVO.tagVO.name ?? ""
+        }
+        
+        let addedNames: [String] = zip(forAddingBool, forAddingNames).filter{ $0.0 }.map{ $1 }
+        
+        dispatchGroup.enter()
+        viewModel?.addTag(tagNames: addedNames, completion: { (result) in
+            dispatchGroup.leave()
+        })
+        
+        let forRemovalBool = sortedArray.map { (item) -> Bool in
+            return item.forRemoval
+        }
+        let forRemovalTags = sortedArray.map { (item) -> TagVO in
+            return item.tagVO
+        }
+        let removedTags: [TagVO] = zip(forRemovalBool, forRemovalTags).filter{ $0.0 }.map{ $1 }
+        
+        dispatchGroup.enter()
+        viewModel?.deleteTag(tagVO: removedTags, completion: { (result) in
+            dispatchGroup.leave()
+            })   
+       
+        dispatchGroup.notify(queue: .main) {
+            self.hideSpinner()
+            self.delegate?.tagDetailsViewControllerDidUpdate(self)
+            self.dismiss(animated: true, completion: nil)
+        }
+
+    }
+    
+    @IBAction func AddTagButtonAction(_ sender: Any) {
+        var tagsList: [String] = []
+        
+        for item in sortedArray {
+            if let itemName = item.tagVO.tagVO.name {
+                tagsList.append(itemName)
+            }
+        }
+        
+        if let findTag = tagFindSearchBar.text,
+           !findTag.isEmpty,
+           !tagsList.contains(findTag)
+        {
+            sortedArray.insert(SortedTagVO(tagVO: TagVO(tagVO: TagVOData(name: findTag, status: String(), tagId: Int(), type: String(), createdDT: String(), updatedDT: String())), checked: true, forAdding: true, forRemoval: false), at: 0)
+            
+            self.tagFindSearchBar.text = ""
+            
+            filteredTagVO = sortedArray.map({ (item) -> TagVO in
+                return item.tagVO
+            })
+            
+            self.tagsCollectionView.reloadData()
+            self.view.endEditing(true)
+        }
+    }
+    
+    func initTagsState()
+    {
+        showSpinner(colored: .lightGray)
+        viewModel?.getTagsByArchive(archiveId: viewModel?.file.archiveId ?? 0, completion: { result in
+            guard let tagsArchive = result else {
+                return
+            }
+            
+            for (index,_) in tagsArchive.enumerated() {
+                self.sortedArray.append(SortedTagVO(tagVO: tagsArchive[index]))
+                
+            }
+            
+            if let tags = self.tagVOS {
```

Comment:
tagVOS property is never written to, is this check and subsequent logic necessary here?

Reply:
tagVOS is passed in from the previous screen, but it's the same as self.recordVO?.tagVOS.
My suggestion is to take out tagVOS and use self.recordVO?.tagVOS instead

Reply:
Removed tagVOS and used recordVO?.tagVOS